packages/core/src/storages/sitemap_request_list.ts

```diff
@@ -0,0 +1,292 @@
+ */
+export class SitemapRequestList implements IRequestList {
+    /**
+     * Set of URLs that were returned by fetchNextRequest() and not marked as handled yet.
+     * @internal */
+    inProgress = new Set<string>();
+
+    /** Set of URLs for which reclaimRequest() was called. */
+    private reclaimed = new Set<string>();
+
+    /** Request objects for in-progress and reclaimed URLs so that we can keep track of state */
+    private requestData = new Map<string, Request>();
+
+    /**
+     * URLs of
+     * 1. sitemaps that have been completely processed and
+     * 2. processed sitemap items of the current (partially processed) sitemap
+     */
+    private handledUrls = { sitemapUrls: new Set<string>(), urls: new Set<string>() };
+
+    /** URLs loaded from the sitemaps */
+    private queuedUrlsBySitemap = new Map<string, string[]>();
+
+    /** Number of URLs that were marked as handled */
+    private handledUrlCount = 0;
+
+    /** Indicates whether the background processing of sitemap contents has already finished.  */
+    private isSitemapFullyLoaded = false;
+
+    private persistStateKey?: string;
+
+    private store?: KeyValueStore;
+
+    private sitemapUrls: string[];
+
+    private proxyUrl: string | undefined;
+
+    /** @internal */
+    private constructor(options: SitemapRequestListOptions) {
+        ow(
+            options,
+            ow.object.exactShape({
+                sitemapUrls: ow.array,
+                proxyUrl: ow.optional.string,
+                persistStateKey: ow.optional.string,
+            }),
+        );
+
+        this.persistStateKey = options.persistStateKey;
+        this.sitemapUrls = options.sitemapUrls;
+        this.proxyUrl = options.proxyUrl;
+    }
+
+    private startLoadingInBackground(): void {
+        (async () => {
+            for await (const item of parseSitemap(
+                this.sitemapUrls.map((url) => ({ type: 'url', url })),
+                this.proxyUrl,
+            )) {
+                if (this.handledUrls.sitemapUrls.has(item.originSitemapUrl)) {
+                    continue;
+                }
+
+                if (!this.queuedUrlsBySitemap.has(item.originSitemapUrl)) {
+                    this.queuedUrlsBySitemap.set(item.originSitemapUrl, []);
+                }
+
+                const queue = this.queuedUrlsBySitemap.get(item.originSitemapUrl)!;
+
+                if (!this.handledUrls.urls.has(item.url)) {
+                    queue.push(item.url);
+                }
+            }
+        })()
+            .then(() => {
+                this.isSitemapFullyLoaded = true;
+            })
+            .catch(() => {
+                this.isSitemapFullyLoaded = true;
+            });
+    }
+
+    /**
+     * Open a sitemap and start processing it.
+     */
+    static async open(options: SitemapRequestListOptions): Promise<SitemapRequestList> {
```

Comment:
@barjin will this work OK for WCC? I'll gladly adjust it if needed.

Reply:
Sure, the API usage in the tests looks comfortable enough to me 👍🏽

Reply:
we should add a timeout to `Sitemap.load`